Resurrect paper (script)

Prof-Bloodstone opened the pull request:
This is a rewrite of the original script, while trying to be as
compatible as possible. There's a few changes though:
1. Patching and path rebuilding commands accept argument to only patch
   api or server.
2. testserver command no longer setups whole server with plugins,
   instead uses runserver config. It also accept specifying whether to
   run dev server (default), shadow or reobfuscated one.

This script now has all the logic in functions, which shifts the burden
of unregistering variables onto the shell - we only need to remember to
cleanup helper functions. Allows for more often use of meaningful
variables, which should hopefully make it easier to read.

I tried to make it as compatible and portable as possible, assuming
basic BASH and ZSH features, but it's very likely that I missed
something.

A few things to consider:
1. What should `build` command look like? There seems to be no difference in the old script, apart from not installing server into maven local: https://github.com/PaperMC/Paper/blob/ver/1.16.5/paper#L62-L82
2. Should (most) commands warn that one should use gradle commands now, instead?
3. What dependencies do we really require - I believe curl is redundant now, but haven't had a chance to test rest yet.

This is a DRAFT as it depends on changes from https://github.com/PaperMC/Paper/pull/5957

CC:
@Proximyst for script review
@electronicboy for MacOS compat check

Prof-Bloodstone committed:
Resurrect paper (script)

Bring back our beloved build tool, better than ever!

This is a rewrite of the original script, while trying to be as
compatible as possible. There's a few changes though:
1. Patching and path rebuilding commands accept argument to only patch
   api or server.
2. testserver command no longer setups whole server with plugins,
   instead uses runserver config. It also accept specifying whether to
   run dev server (default), shadow or reobfuscated one.

This script now has all the logic in functions, which shifts the burden
of unregistering variables onto the shell - we only need to remember to
cleanup helper functions. Allows for more often use of meaningful
variables, which should hopefully make it easier to read.

I tried to make it as compatible and portable as possible, assuming
basic BASH and ZSH features, but it's very likely that I missed
something.

Prof-Bloodstone committed:
Fix ZSH script path extraction

Returning an absolute path to a function inside a script is so dum

kashike commented on paper:
```diff
@@ -0,0 +1,298 @@
+  # TODO: Verify which ones are required
+  _paper_is_dep_available git
+  _paper_is_dep_available patch
+  _paper_is_dep_available mvn
```
nope

Prof-Bloodstone replied:
Confirmed unnecessary and removed.

Proximyst commented on paper:
```diff
@@ -0,0 +1,298 @@
+
+# Can't use `set -eu` since ppl might source this file
+
+# TODO: Display warning that one should use gradle?
```
Wrap it in a subshell, then. It's stupid, but whatever.

Prof-Bloodstone replied:
I'm not sure how it applies here? This is my second point from things to consider in the PR description:
> 2. Should (most) commands warn that one should use gradle commands now, instead?

Do we want to let PPL know that gradle replaced (most of) this script and they can use it directly for most things? Probably with an env variable to suppress it.

Prof-Bloodstone replied:
Oh, I feel like your comment might have been to that third line - can't do that, since we can't use subshells for the whole scripts due to us wanting to be able to change user CWD after some of the commands.

DenWav commented on paper:
```diff
@@ -0,0 +1,298 @@
+  _paper_is_dep_available mvn
+  _paper_is_dep_available curl
+  _paper_is_dep_available javac "${java_not_found_message}"
+  _paper_is_dep_available jar "${java_not_found_message}"
```
not anymore (curl, javac, jar)

Prof-Bloodstone replied:
I confirmed curl is only used for checkoutpr script, which isn't used by this script, so it's fine to remove. But you still need the JDK - `gradlew` will show a decent message when JRE is not installed, but when JDK is missing, it shows this exception:
```
> Task :Paper-API:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':Paper-API:compileJava'.
> Cannot invoke "javax.tools.JavaCompiler.getStandardFileManager(javax.tools.DiagnosticListener, java.util.Locale, java.nio.charset.Charset)" because "this.delegate" is null

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
```
Maybe it's worth leaving them as means for JDK detection?

electronicboy replied:
I was thinking that the JDK missing might actually produce a nicer message, I mean, in part it kinda makes sense, the JavaCompiler is not there, so, probs JDK specific but, erm... normies. as said before, these scripts should generally serve as idiot checks but don't need to be fool proof, the notation of ensuring that specific binaries are on the path is pretty gross especially when we don't *need* them to be on the path itself is pretty meh, but, for the sake of idiot checks? eeeer... am not sure

DenWav replied:
You don't need a JDK, no. If it's erroring then I messed something up, will check on it.

DenWav replied:
Also (lol):

> eeeer... am not sure

I am sure. A properly set up project does not need a JDK. Gradle handles that: https://github.com/PaperMC/Paper/blob/0ad55267a387bf9cea3b6d0caeb843b3253cca65/build.gradle.kts#L12-L14

DenWav replied:
Can't reproduce, builds fine without a JDK anywhere to be found.

Prof-Bloodstone replied:
Reproduced on a fresh machine:
Order CX31 from Hetzner (cheapest 8GB machine), with Debian 10 and your SSH key.

Then run the following:
```sh
apt update
# Setup git
apt install git
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
# Setup Java (paper documentation for Debian)
apt install apt-transport-https software-properties-common gnupg wget
wget -qO - https://adoptopenjdk.jfrog.io/adoptopenjdk/api/gpg/key/public | apt-key add -
add-apt-repository https://adoptopenjdk.jfrog.io/adoptopenjdk/deb/
apt update
apt install adoptopenjdk-16-hotspot-jre # Install JRE, not JDK
# Clone paper and build
git clone https://github.com/PaperMC/Paper.git
cd Paper
./gradlew applyPatches
./gradlew paperclipJar
```
Result of that last command:
```
root@paper-test:~/Paper# ./gradlew paperclipJar
> Task :Paper-API:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':Paper-API:compileJava'.
> Cannot invoke "javax.tools.JavaCompiler.getStandardFileManager(javax.tools.DiagnosticListener, java.util.Locale, java.nio.charset.Charset)" because "this.delegate" is null

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 20s
18 actionable tasks: 5 executed, 13 up-to-date
```

It somehow needs to have a JDK (because it has to compile that gradle code from Kotlin, right?), but fails when I don't install full JDK.

I'm fine with removing it despite that.

DenWav replied:
I did test with a fresh machine with only a JRE lol.

The failure is literally an NPE:
> Cannot invoke "javax.tools.JavaCompiler.getStandardFileManager(javax.tools.DiagnosticListener, java.util.Locale, java.nio.charset.Charset)" because "this.delegate" is null

I don't doubt the failure, but I'll follow your repro steps to figure out what's really going on.

DenWav commented on paper:
```diff
@@ -0,0 +1,298 @@
+  local -r java_not_found_message="was not found - download the JDK by following https://paper.readthedocs.io/en/latest/java-update/index.html"
+  # TODO: Verify which ones are required
+  _paper_is_dep_available git
+  _paper_is_dep_available patch
```
not anymore (patch)

Prof-Bloodstone replied:
Confirmed unnecessary and removed.

DenWav commented on paper:
```diff
@@ -0,0 +1,298 @@
+      (
+        cd "${basedir}"
+        ./gradlew applyPatches || return "${?}"
+        ./gradlew paperclipJar || return "${?}"
```
There's paperclip command below, I think this should be `reobfJar` instead.

Prof-Bloodstone replied:
Yeah, it was based on the previous script: https://github.com/PaperMC/Paper/blob/ver/1.16.5/paper#L55-L61
I'll make it build `reobfJar` instead.

DenWav commented on paper:
```diff
@@ -0,0 +1,298 @@
+      );;
+    b|build)
+      (
+      printf 'NOT IMPLEMENTED YET\n' >&2 # TODO
```
This shouldn't be any more complicated than `./gradlew build`

Prof-Bloodstone replied:
Done.

DenWav commented on paper:
```diff
@@ -0,0 +1,298 @@
+      (
+        cd "${basedir}"
+        ./gradlew makeMcDevSrc || return "${?}"
+      );;
```
This should be removed. I will be removing the task as soon as I find a proper way to implement this.

As it currently is, due to how IntelliJ handles interfacing with Gradle, this task does nothing.

Prof-Bloodstone replied:
Removed.

DenWav commented on paper:
```diff
@@ -0,0 +1,298 @@
+        # Is this good or should be OG version?
+        ./gradlew clean cleanCache || return "${?}"
+      );;
+    con|continue)
+        if [ -d ".git/rebase-apply" ]; then
+            _papergit am --continue
+        elif [ -d ".git/rebase-merge" ]; then
+            _papergit rebase --continue
+        fi
+      ;;
+    e|edit)
```
Would love if we could get rid of these entirely tbh.

These can cause issues esp when used by someone who doesn't know what they're doing. I think it's best to just not have it at all and have people do this themselves (it's not complicated).

But if there's pushback then sure, leave it.

electronicboy replied:
The way this command worked was always such a clusterfuck that if it disappeared, I really wouldn't give a single hoot, edit could maybe exist and just run the command for you, but, otherwise, the continue commands definitely were always a pain in the ass, you couldn't edit API and server at the same time and continue, and in general it was just kinda cludgy

Prof-Bloodstone replied:
If the issue is that you can't edit server and api at the same time, I believe it can be fixed.

I can see the value in keeping this, but fine if you want to remove it.

electronicboy replied:
I'm not exactly for or against it, but, I mean, the way continue works means that you can't edit both at the same time as it forgets the last one, it's tripped so many people up over the years with how it works, I'd maybe keep the edit stuff just for laziness, but, am not really sure about `edit continue`

Prof-Bloodstone replied:
What if we make it so you can do:

```sh
$ paper edit api
$ paper edit server
$ paper edit continue # Will continue in server
$ paper edit continue
No prior edit work to continue - try explicitly passing "api" or "server"?
$ paper edit continue api
```

Other case:

```sh
$ paper edit api
$ paper edit server
$ paper edit continue api
$ paper edit continue # Will continue server? Though I feel like API would be more intuitive?
```

I feel like it's not a very straightforward command either way, and to use it you need some understanding and decent CONTRIBUTING docs.

The more I look at it, the more confusing it becomes :smiley:

electronicboy replied:
That's the issue, the more you toy with it the more of a clusterfuck it becomes to the degree that I don't entirely see the merit in edit continue

DenWav replied:
That CLI interface looks really confusing....I don't think this is benefiting anyone.

Prof-Bloodstone replied:
Removed `edit continue`, left `edit api` and `edit server`.

Phoenix616 replied:
Why not do something like this: `paper edit api continue` and `paper edit server continue` (that's what [I did](https://github.com/Minebench/Origami/blob/1.16/origami#L56-L76) in my fork script), that way one can execute that extremely fast with the sequence `UP key` `SPACE` `c` `ENTER` while still being able to specify what to continue without manually doing it inside the folders.

Not having a continue function there would only resolve half of the complexity of the git actions, having a dedicated continue argument per type would fix the old issues with editing both at the same time as well as it sometimes not properly storing the edited type for the automatic continue.

DenWav reviewed (changes requested):
Some things - some of them are opinions, others aren't. Hopefully I made it clear which ones are opinions. Feel free to ignore those.

Prof-Bloodstone committed:
Review changes

Prof-Bloodstone committed:
Remove last TODO

Prof-Bloodstone committed:
Remove mcdev and edit continue

Prof-Bloodstone committed:
Make paper script work on bash 3.2.57

Support the poor souls stuck using it

Apparently there's a bug with 'command' immediately exiting whole script on error...

Prof-Bloodstone committed:
Remove paperclip and install commands
